Public/Src/Cache/ContentStore/Distributed/NuCache/FlushableCache.cs

```diff
@@ -0,0 +1,165 @@
+            }
+        }
+
+        /// <nodoc />
+        public bool TryGetEntry(ShortHash hash, out ContentLocationEntry entry)
+        {
+            using (_exchangeLock.AcquireReadLock())
+            {
+                // The entry could be a tombstone, so we need to make sure the user knows content has actually been
+                // deleted, which is why we check for null.
+                if (_cache.TryGetValue(hash, out entry))
+                {
+                    _database.Counters[ContentLocationDatabaseCounters.TotalNumberOfCacheHit].Increment();
+                    return entry != null;
+                }
+                else if (_flushingCache.TryGetValue(hash, out entry))
+                {
+                    _database.Counters[ContentLocationDatabaseCounters.TotalNumberOfCacheHit].Increment();
+                    return entry != null;
+                }
+            }
+
+            _database.Counters[ContentLocationDatabaseCounters.TotalNumberOfCacheMiss].Increment();
+
+            return false;
+        }
+
+        /// <nodoc />
+        public async Task FlushAsync(OperationContext context)
+        {
+            // This lock is required to ensure no flushes happen concurrently. We may loose updates if that happens.
+            // AcquireAsync is used so as to avoid multiple concurrent tasks just waiting; this way we return the
+            // task to the thread pool in between.
+            using (await _flushTurnstile.AcquireAsync())
+            {
+                PerformFlush(context);
+            }
+        }
+
+        private void PerformFlush(OperationContext context)
+        {
+            _database.Counters[ContentLocationDatabaseCounters.TotalNumberOfCacheFlushes].Increment();
+
+            using (_database.Counters[ContentLocationDatabaseCounters.CacheFlush].Start())
+            {
+                Contract.Assert(_flushingCache.Count == 0);
+
+                using (_exchangeLock.AcquireWriteLock())
+                {
+                    _flushingCache = _cache;
+                    _cache = new ConcurrentBigMap<ShortHash, ContentLocationEntry>();
+                }
+
+                if (_configuration.FlushSingleTransaction)
+                {
+                    _database.PersistBatch(context, _flushingCache);
+                }
+                else
+                {
+                    var actionBlock = new ActionBlockSlim<KeyValuePair<ShortHash, ContentLocationEntry>>(_configuration.FlushDegreeOfParallelism, kv =>
+                    {
+                        // Do not lock on GetLock here, as it will cause a deadlock with
+                        // SetMachineExistenceAndUpdateDatabase. It is correct not do take any locks as well, because
+                        // no Store can happen while flush is running.
+                        _database.Persist(context, kv.Key, kv.Value);
+                    });
+
+                    foreach (var kv in _flushingCache)

```

Comment:
>_flushingCache [](start = 39, length = 14)

In theory you could call GetPages() in to create batches rather than using a single item database.Persist()

Reply:
Indeed, I'd rather not do this change in this PR for now, but in a follow-up if we end up seeing that single-threaded flushing of WriteBatches is indeed problematic

---
In reply to: [292139868](https://github.com/microsoft/BuildXL/pull/403#discussion_r292139868) [](ancestors = 292139868)